Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs

```diff
@@ -433,57 +456,76 @@ public ActionResult DeleteAll(int? id)
             }
 
             if (!this.CheckIfUserHasContestPermissions(id.Value))
             {
                 this.TempData.AddDangerMessage(GlobalConstants.NoPrivilegesMessage);
-                return this.RedirectToAction("Index", "Contests", new { area = "Administration" });
+                return this.RedirectToAction<ContestsController>(c => c.Index());
             }
 
-            var contest = this.Data.Contests.All()
-                .Where(x => x.Id == id)
-                .Select(ContestAdministrationViewModel.ViewModel)
-                .FirstOrDefault();
+            var contest = this.contestsData.GetById(id.Value);
 
             if (contest == null)
             {
                 this.TempData.AddDangerMessage(GlobalResource.Invalid_contest);
-                return this.RedirectToAction(nameof(this.Index));
+                return this.RedirectToAction(c => c.Index());
+            }
+
+            if (contest.CanBeCompeted)
+            {
+                this.TempData.AddDangerMessage(GlobalResource.Active_contest_problems_permitted_for_deletion);
+                return this.RedirectToAction(c => c.Contest(id.Value));
             }
 
-            return this.View(contest);
+            var contestModel = new DeleteProblemsFromContestViewModel
+            {
+                Id = contest.Id,
+                Name = contest.Name
+            };
+
+            return this.View(contestModel);
         }
 
-        public ActionResult ConfirmDeleteAll(int? id)
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public ActionResult ConfirmDeleteAll(int contestId)
         {
-            if (id == null || !this.Data.Contests.All().Any(x => x.Id == id))
+            if (!this.CheckIfUserHasContestPermissions(contestId))
             {
                 this.TempData.AddDangerMessage(GlobalResource.Invalid_contest);
-                return this.RedirectToAction<ProblemsController>(x => x.Index());
+                return this.RedirectToAction(c => c.Index());
             }
 
-            if (!this.CheckIfUserHasContestPermissions(id.Value))
+            var contest = this.contestsData.GetById(contestId);
+
+            if (contest == null)
             {
                 this.TempData.AddDangerMessage(GlobalResource.Invalid_contest);
-                return this.RedirectToAction(nameof(this.Index));
+                return this.RedirectToAction(c => c.Index());
             }
 
-            this.Data.Resources.Delete(r => r.Problem.ContestId == id);
+            if (contest.CanBeCompeted)
+            {
+                this.TempData.AddDangerMessage(GlobalResource.Active_contest_problems_permitted_for_deletion);
+                return this.RedirectToAction(c => c.Contest(contest.Id));
+            }
 
-            this.Data.TestRuns.Delete(tr => tr.Submission.Problem.ContestId == id);
+            this.Data.Resources.Delete(r => r.Problem.ContestId == contestId);
```

Comment:
Why call the Resources repository instead of the Problems one?

Reply:
I call the problemsDataService for the deletion of the problem